README.md
## A Async Template for Node.js, Use ejs template language [![Build Status](https://travis-ci.org/DoubleSpout/ariestp.svg?branch=master)](https://travis-ci.org/DoubleSpout/ariestp)

##Install

		npm install ariestp

###SimpleExample

		"use strict";
		const aries = require("ariestp");
		const fs = require("fs");
		const http = require("http");
		
		const asyncTpl = `<%?
			ctx.thinking;
			setTimeout(function () {
			  ctx.thinking = "1 seconde later"
			  aries();      
			}, 1000);
			
			%>
			<%- ctx.thinking %>
			
			<%?
			ctx.http.get('http://www.baidu.com/', (res) => {
				ctx.status =  res.statusCode;
				res.resume();
				aries();
			}).on('error', (e) => {
			  ctx.status = 500;
			  aries();
			});
			%>
			<%- ctx.status %>
			`;
		
		aries.compile(asyncTpl, {http:http}, (err, renderStr, isUseCache) => {
				if(err) return console.log(err);
				console.log(renderStr);
		});
	   

##API
   
### aries.set(opt)
	
	path: __dirname, 	// template find root path
	includeId: function,		// ex:  (templateId, cb) => return fs.readFile(path.join(__dirname,templateId+'.html'), cb);
	cacheTime:10000,	// 编译缓存时间，单位：毫秒

### aries.compileFile(filepath, ctx, cb)
	
	filepath:"async_example.html"	// template relate to root path
	ctx: {}							// template ctx object
	cb: function					// ex:  (err, renderStr, isUseCache) => {}

	
### aries.compile(templateStr, ctx, cb)
	
	templateStr:"<%= ctx.xxxx %>"	// template relate to root path
	ctx: {}							// template ctx object
	cb: function					// ex:  (err, renderStr, isUseCache) => {}



## Template Language

### async tag

		<%?   
			setTimeout(function () {
				  ctx.thinking = "1 seconde later"
				  aries();  //make sure to call aries(); function to end the async function.
			}, 1000);
		%>
		

### example
	
see more example in `example` folder
	
		<!DOCTYPE html>
		<html lang="zh_CN" class="html-">
		<body>
			<!-- test let fature -->	
			<% let testLet = "let"  %>
			<h4>aries is support that: <%= testLet %></h4>

			<!-- test const fature	 -->
			<% const testConst = "const";  %>
			<h4>aries is support that: <%= testConst %></h4>

			<!-- test template string fature -->	
			<% let tempString = "string"; %>
			<% let templateString = `template ${tempString} `;%>
			<h4>aries is support that: <%= templateString %></h4>	

			<!-- test string.raw string fature -->	
			<% var stringRaw = String.raw({ raw: 'srn.a ' }, 't', 'i', 'g', 'r', 'w'); %>
			<h4>aries is support that: <%= stringRaw %></h4>	

			<!-- test Symbol fature -->	
			<% if( Symbol.for("bar") === Symbol.for("bar") ){ %>
				<h4>aries is support that: <%= "Symbol" %></h4>	
			<% } %>

			<!-- test map fature -->	
			<% var m = new Map(); %>
			<% var o = {p: "Hello World"};%>
			<% m.set(o, "map"); %>
			<% let testMap = m.get(o); %>
			<h4>aries is support that: <%= testMap %></h4>	

			<!-- <%= ctx.fsdsdfsdfsd %> -->	
			<% function* helloWorldGenerator() { %>
				<% yield 'Generator'; %>
				<% return 'ending'; %>
			<% } %>
			<% let hw = helloWorldGenerator(); %>
			<% let testGenerator = hw.next().value; %>
			<h4>aries is support that: <%= testGenerator %></h4>	

			<!-- test arrow functions fature -->	
			<% let testArrowFunction = ""; %>
			<% ["arrow","fuction"].forEach((x) => { %>
				<% testArrowFunction += ' ' +x; %>
			<% })%>
			<h4>aries is support that: <%= testArrowFunction %></h4>
			
			<!-- test for .. of fature -->
			<% var str = ''; %>
			<% for (let item of 'for...of') { %>
				<% str += item; %>
			<% } %>
			<h4>aries is support that: <%= str %></h4>
			
			<!-- test for .. of fature -->
			<% include "foot.html" %>
			
			
			<!-- test for .. of fature ,includeId and transfer param, in foot tpl, use ctx.param1 and ctx.param2 to get it-->
			<% includeId "foot?param1=1&param2=2" %>
			
		</body>
		</html>

	
## MIT

```
Copyright (c) 2013 wu zhonghua <snoopyxdy@gmail.com>

The MIT License

Permission is hereby granted, free of charge, to any person obtaining
a copy of this software and associated documentation files (the
"Software"), to deal in the Software without restriction, including
without limitation the rights to use, copy, modify, merge, publish,
distribute, sublicense, and/or sell copies of the Software, and to
permit persons to whom the Software is furnished to do so, subject to
the following conditions:

The above copyright notice and this permission notice shall be
included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE
LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
```

Changes.md
# Changes

This file is a manually maintained list of changes for each release. Feel free
to add your changes here when sending pull requests. Also send corrections if
you spot any mistakes.

## v0.1.8 (2016-04-07)
*注释删除
*自动分号


## v0.1.7 (2016-04-07)
*可以在模板里使用es6函数

## v0.1.6 (2016-04-06)
*增加转义功能

## v0.1.3 (2016-03-30)
* publish version



test/test.js
"use strict";
const aries = require("../index.js");
const assert = require('assert');
const async = require('async');
const fs = require('fs');
const path = require('path');
//设置白羊座
aries.set({
	path:__dirname,
	includeId:function(templateId, cb){
		return fs.readFile(path.join(__dirname,templateId+'.html'), cb)
	},
	cacheTime:10000,
});

let taskList = [];
let taskList2 = [];


// 测试普通字符串
taskList.push(function(callback){
	let syncStr = "<html><body>helloworld</body></html>";
	aries.compile(syncStr, {}, (err, renderStr, isUseCache) => {
		   if(err){
				console.log(err.stack)
		   }
		   assert(!err);
		   assert(!isUseCache);
		   assert(renderStr === "<html><body>helloworld</body></html>");

		   aries.compile(syncStr, {}, (err, renderStr, isUseCache) => {
			   assert(!err);
			   assert(isUseCache);
			   assert(renderStr === "<html><body>helloworld</body></html>");
			   console.log("normal str test ok");
			   callback();
			})
	})
});


// //测试转移字符串
taskList.push(function(callback){
	let syncStr = "<%= ctx.data %>";
	aries.compile(syncStr, {data:"<p>\"\'</p>"}, (err, renderStr, isUseCache) => {
			if(err){
				 console.log(err.stack)
			}
			console.log(renderStr)
		   assert(!err);
		   assert(!isUseCache);
		   assert(renderStr === "&lt;p&gt;&quot;&apos;&lt;/p&gt;");

		   aries.compile(syncStr, {data:"<p>\"\'</p>"}, (err, renderStr, isUseCache) => {
			   assert(!err);
			   assert(isUseCache);
			   assert(renderStr === "&lt;p&gt;&quot;&apos;&lt;/p&gt;");
			   console.log("escape sync str test ok");
			   callback();
			})
	})
});


// //测试渲染字符串
taskList.push(function(callback){
	let syncStr = "<%- ctx.data %>";
	aries.compile(syncStr, {data:"123"}, (err, renderStr, isUseCache) => {
			if(err){
				 console.log(err.stack)
			}
		   assert(!err);
		   assert(!isUseCache);
		   assert(renderStr === "123");

		   aries.compile(syncStr, {data:"1234"}, (err, renderStr, isUseCache) => {
			   assert(!err);
			   assert(isUseCache);
			   assert(renderStr === "1234");
			   console.log("sync str test ok");
			   callback();
			})
	})
});


// //测试渲染异步模板字符串
taskList.push(function(callback){
	let syncStr = "<%? setTimeout(function(){ctx.data='123';aries();},ctx.timeout) %><%- ctx.data %>";
	aries.compile(syncStr, {timeout:1000}, function(err, renderStr, isUseCache){
			if(err) {
				 console.log(err.stack)
			}
		   assert(!err);
		   assert(!isUseCache);
		   assert(renderStr === "123");

		   aries.compile(syncStr, {timeout:1000}, (err, renderStr, isUseCache) => {
			   assert(!err);
			   assert(isUseCache);
			   assert(renderStr === "123");	
			   console.log("async str test ok");
			   		callback();
			})
	})
});



// //测试同步include文件功能
taskList.push(function(callback){
	let syncStr = "<% include sync.html %>";
	aries.compile(syncStr, {data:"123"}, (err, renderStr, isUseCache, includeIds, includeFs) => {
		   
		   assert(!err);
		   assert(!isUseCache);
		   assert(renderStr === "123");
		   assert(includeFs.length === 1)
		   assert(includeFs[0] === "sync.html")


		   aries.compile(syncStr, {data:"1234"}, (err, renderStr, isUseCache) => {
			   assert(!err);
			   assert(isUseCache);
			   assert(renderStr === "1234");	
			   console.log("include test ok");		
			   callback();
			})
	})
});


// //测试异步includeId文件功能
taskList.push(function(callback){
	let syncStr = "<% includeId async %>";
	aries.compile(syncStr, {data:"123"}, function(err, renderStr, isUseCache, includeIds, includeFs){
			
		   assert(!err);
		   assert(!isUseCache);
		   assert(renderStr === "123");
		   
		   assert(includeIds.length === 1)
		   assert(includeIds[0] === "async")

		   aries.compile(syncStr, {data:"123"}, (err, renderStr, isUseCache) => {
			   assert(!err);
			   assert(isUseCache);
			   assert(renderStr === "123");	
			   console.log("includeId test ok");		
			   callback();
			})
	})
});


// // //测试include嵌套
taskList.push(function(callback){
	let syncStr = "<% include incLv1.html %>";
	aries.compile(syncStr, {data:"123"}, (err, renderStr, isUseCache, includeIds, includeFs) => {
			if(err){
				console.log(err.stack)
			}
		   assert(!err);
		   assert(!isUseCache);
		   assert(renderStr === "123");

		   assert(includeFs.length === 4)
		   assert(includeFs[0] === "incLv1.html")
		   assert(includeFs[1] === "incLv2.html")
		   assert(includeFs[2] === "incLv3.html")
		   assert(includeFs[3] === "incLv4.html")

		  
		   aries.compile(syncStr, {data:"123"}, (err, renderStr, isUseCache) => {
			   assert(!err);
			   assert(isUseCache);
			   assert(renderStr === "123");	
			   console.log("include embed test ok");		
			   callback();
			})
	})
});



// // //测试include嵌套
taskList.push(function(callback){
	let syncStr = "<% include incLv1.html %>";
	
	aries.includes(syncStr, (err, includeIds, includeFs) => {
		   if(err){
				console.log(err)
		   }
		   assert(!err);

		   assert(includeFs.length === 4)
		   assert(includeFs[0] === "incLv1.html")
		   assert(includeFs[1] === "incLv2.html")
		   assert(includeFs[2] === "incLv3.html")
		   assert(includeFs[3] === "incLv4.html")

		   console.log("include embed only include test ok");	
		   callback();
		
	})
});


// //测试 if else 语句
taskList.push(function(callback){
	let syncStr = "<% if(false){ %><%- ctx.data %><%} else {%><%- ctx.data %><%}%>";
	aries.compile(syncStr, {data:"123"}, (err, renderStr, isUseCache) => {
		   assert(!err);
		   assert(!isUseCache);
		   assert(renderStr === "123");

		   aries.compile(syncStr, {data:"1234"}, (err, renderStr, isUseCache) => {
			   assert(!err);
			   assert(isUseCache);
			   assert(renderStr === "1234");
			   console.log("if else test ok");
			   callback();
			})
	})
});

// //测试 for 语句
taskList.push(function(callback){
	let syncStr = "<% for(var i=1;i<ctx.data;i++){%><%- i %><%}%>";
	aries.compile(syncStr, {data:10}, (err, renderStr, isUseCache) => {
		   assert(!err);
		   assert(!isUseCache);
		   assert(renderStr === "123456789");

		   aries.compile(syncStr, {data:5}, (err, renderStr, isUseCache) => {
			   assert(!err);
			   assert(isUseCache);
			   assert(renderStr === "1234");
			   console.log("for loop test ok");		
			   callback();
			})
	})
});




//测试获取模版文件，相对路径
taskList.push(function(callback){
	let tplPath = "async/async.html";
	aries.compileFile(tplPath, {data:10}, (err, renderStr, isUseCache) => {
		   if(err){
		   		console.log(err.stack)
		   }
		   assert(!err);
		   assert(!isUseCache);
		   assert(renderStr === "20");

		   aries.compileFile(tplPath, {data:20}, (err, renderStr, isUseCache) => {
			   assert(!err);
			   assert(isUseCache);
			   assert(renderStr === "40");
			   console.log("tpl async tpl test ok");		
			   callback();
			})
	})
});


//测试获取模版文件，绝对路径
taskList.push(function(callback){
	let tplPath = path.join(__dirname,"async", "async.html");
	console.log("tplPath abs path "+tplPath)
	aries.compileFile(tplPath, {data:10}, (err, renderStr, isUseCache) => {
		   assert(!err);
		   assert(isUseCache);//已经缓存
		   assert(renderStr === "20");

		   aries.compileFile(tplPath, {data:20}, (err, renderStr, isUseCache) => {
			   assert(!err);
			   assert(isUseCache);
			   assert(renderStr === "40");
			   console.log("tpl abs path tpl test ok");		
			   callback();
			})
	})
});

//测试include模板传参
taskList.push(function(callback){
	let syncStr = "<% ctx.data = 1%><% includeId paramLv1?data=a %>|<%= ctx.data %>";
	aries.compile(syncStr, {}, function(err, renderStr, isUseCache, includeIds, includeFs){
			
		   assert(!err);
		   assert(!isUseCache);
		   assert(renderStr.trim() === "a|a");
		   console.log("tpl includeId param test ok");		
		   callback();
	})
});


//测试includefile模板传参
taskList.push(function(callback){
	let syncStr = "<% ctx.data = 1%><% include paramLv1.html?data=a %>|<%= ctx.data %>";
	aries.compile(syncStr, {}, function(err, renderStr, isUseCache, includeIds, includeFs){
			
		   assert(!err);
		   assert(!isUseCache);
		   assert(renderStr.trim() === "a|a");
		   console.log("tpl include file param test ok");		
		   callback();
	})
});



async.series(taskList, function(err){
	if(err){
		console.log(err);
		return
	}
	console.log("all test ok");
	process.exit(0)
})

lib/aries.js
var fs = require('fs');
var path = require('path');
var url = require('url');
var util = require('util');
var md5 = require('blueimp-md5');
var ariestp = require("./ariestp.js")
var te = ariestp.Engine();
var async = require('async');
var domain = require('domain');
var hasSet = false;
var atry = require('./atry.js');
var aries = {};


aries.set = function(opt){
	if(hasSet) return te;
	
	if(opt.path){
		//检查 opt.path是否存在
		fs.exists(opt.path, function (exists) {
		  if(!exists){
			 var errMsg = "[ariestp] ariestp.set opt.path is not exist!"
			 console.error(errMsg);
			 throw new Error(errMsg);
		  }
		});
	}
	
	if(opt.includeId && "function" !== typeof opt.includeId){
		var errMsg = "[ariestp] ariestp.set opt.includeId must be function!"
		console.error(errMsg);
		throw new Error(errMsg);
	}
	
	if(opt.cacheTime){
		if(isNaN(opt.cacheTime - 0)){
			var errMsg = "[ariestp] ariestp.set opt.cacheTime must be a number!"
			console.error(errMsg);
			throw new Error(errMsg);
		}
		if(opt.cacheTime < 10*1000){
			var errMsg = "[ariestp] ariestp.set opt.cacheTime must gte 10 secs!"
			console.error(errMsg);
			throw new Error(errMsg);
		}
		if(opt.cacheTime > 3600*1000){
			var errMsg = "[ariestp] ariestp.set opt.cacheTime must lt 1 hour!"
			console.error(errMsg);
			throw new Error(errMsg);
		}
		
	}
	
	te = ariestp.Engine(opt);
	hasSet = true;
	console.log("[ariestp] ariestp.set ok, ready for compile.");
	return te;
}



//渲染文件路径
aries.compileFile = function(filename, ctx, cb){
	aries.getTemplateFromFile(filename, function(err, tempstr){
		if(err) return cb(err);
		aries.compile(tempstr.toString(), ctx, cb);
	})
}

//渲染文件字符串
aries.includes = function(tempstr, cb){
	cb.isIncludes = true;
	aries.compile(tempstr, {}, cb);
}

//渲染文件字符串
aries.compile = function(tempstr, ctx, cb){
				
		var compileFinish = false;
		if(!tempstr){
			return cb(null, "");
		}
		var ctx = ctx || {};

		try{
			var key = md5(tempstr);
		}
		catch(e){
			return cb(e);
		}
		var tempStrList = [tempstr];
		
		aries.scanInclude(tempStrList, 0, ctx, function(err, strList, noInclude, includeIds, includeFiles){

				//避免回调2次
				if(compileFinish) return;
				if(err){
					compileFinish = true
					return cb(err);
				}
				
				//如果仅仅返回includeIds
				if(cb.isIncludes){
					return cb(null, includeIds, includeFiles)
				}
				
				//去渲染模版
				var templatePromise = te.template(strList, key, function (err, t, isUseCache) {
						//避免回调2次
						if(compileFinish) return;

						if(err){
							compileFinish = true
							return cb(err);
						}
						var templateResponse = t(te, key, ctx, templatePromise);
						var bodyList = [];

						templateResponse.addListener("body", function (chunk) {
								if(compileFinish) return
								bodyList.push(chunk);
						});

						templateResponse.addListener("complete", function () {
								if(compileFinish) return;
								compileFinish = true;
								//如果出现错误
								if(ctx._ariestpError){
									return cb(ctx._ariestpError, null, isUseCache)
								}
								cb(null, bodyList.join(""), isUseCache, includeIds, includeFiles);
								
						});
						
						templateResponse.addListener("error", function (err) {

								if(compileFinish) return
								compileFinish = true

								cb(err);
						});
				});
		
		});
		
}


aries.getTemplateFromFile = function getTemplateFromFile(filename, cb){
	//是否是绝对路径
	if (filename.charAt(0) === '/' || /^[a-zA-Z]:$/.test(filename.slice(0,2))) {
		filename = filename;
    } else{
		filename = path.join(te.path, filename);
	}
			
	fs.readFile(filename, function (err, tempstr) {
		if (err) return cb(err);
		return cb(null, tempstr)
	});
}


//内部方法扫描是否具有包含include的方法
aries.scanInclude = function scanInclude(tempStrList, count, ctx, cb){
	//符号用
	var templateSymbol = [];
	var symbolPos = -1;	
	var isPairing = false;
	var tmpS, tmpE;
	
	//如果没有就创建数组
	if(typeof cb.includeFiles == "undefined"){
		cb.includeFiles = [];
	}
	
	if(typeof cb.includeIds == "undefined"){
		cb.includeIds = [];
	}
		

	//大于5次的嵌套就出错，防止死循环嵌套
	count++;
	if(count>5){
		var err = new Error(util.format("[ariestp] scanInclude include count lt 5 times"));
		return cb(err);
	}

	//注释用
	// var noteList = [];
	// var notePos = -1;
	// var isPairingNote = false;
	// var tmpNS, tmpNE;
	// 去除注释
	// 字符串长度
	// var len = tempStrList[0].length;
	// for(var i=0;i<len; i++){		
	// 	// <!-- 11916796840964503050040715 -->
	// 	//判断注释
	// 	if(tempStrList[0][i] === '<' && tempStrList[0][i+1] === '!' && tempStrList[0][i+2] === '-' && tempStrList[0][i+3] === '-'){
	// 		if(isPairingNote === true){
	// 			var err = new Error(util.format("[ariestp] template str scan error, not pair <!-- at template char %s, str copy:", i, tempStrList[0].slice(tmpNS, i+4)));
	// 			return cb(err);
	// 		}
	// 		tmpNS = i;
	// 		noteList.push({
	// 			s:i,
	// 			e:-1,
	// 		})
	// 		notePos = noteList.length - 1;		
	// 		isPairingNote = true;

	// 	}else if(tempStrList[0][i] === '-' && tempStrList[0][i+1] === '-' && tempStrList[0][i+2] === '>'){
	// 		if(isPairingNote !== true){
	// 			var err = new Error(util.format("[ariestp] template str scan error, not pair --> at template char %s, str copy:", i, tempStrList[0].slice(tmpNE, i+3)));
	// 			return cb(err);
	// 		}
	// 		tmpE = i + 3;
	// 		noteList[notePos]['e'] = tmpE;
	// 		isPairingNote = false;
	// 		notePos = -1;
	// 	}
	// }

	// //如果有注释
	// if(noteList.length > 0){
	// 	var noteLen = noteList.length;
	// 	//去掉注释
	// 	for(var i=noteLen-1; i>=0;i--){
	// 		tempStrList[0] = tempStrList[0].slice(0, noteList[i]["s"]) + tempStrList[0].slice(noteList[i]["e"]);
	// 	}
	// }

	//字符串长度
	var len = tempStrList[0].length;
	for(var i=0;i<len; i++){
		if(tempStrList[0][i] === '<' && tempStrList[0][i+1] === '%'){
			if(isPairing === true){
				var err = new Error(util.format("[ariestp] template str scan error, not pair <% at template char %s, str copy:", i, tempStrList[0].slice(tmpS, i+2)));
				return cb(err);
			}
			tmpS = i;
			templateSymbol.push({
				s:i,
				e:-1,
			})
			symbolPos = templateSymbol.length - 1;		
			isPairing = true
		} else if(tempStrList[0][i] === '%' && tempStrList[0][i+1] === '>'){
			if(isPairing !== true){
				var err = new Error(util.format("[ariestp] template str scan error, not pair %> at template char %s, str copy:", i, tempStrList[0].slice(tmpE, i+2)));
				return cb(err);
			}
			tmpE = i + 2;
			templateSymbol[symbolPos]['e'] = tmpE;
			isPairing = false;
			symbolPos = -1;
		}		
	}
	
	//所有 <% xxxx %> 的匹配，存入临时数组
	var len = templateSymbol.length;
	var includeList = [];
	var includeReg = /^<%\s+include\s(.+)\s+%>$/;
	var includeIdReg = /^<%\s+includeId\s(.+)\s+%>$/;
	var renderReg = /^<%[=|-](.+)%>$/;
	var hasFh = /^<%(.+);(\s*?)%>$/;

	var needSemicolon = false;
	var include = 0;
	for(var i=0;i<len;i++){
		var includeMatchStr = tempStrList[0].slice(templateSymbol[i]['s'], templateSymbol[i]['e']);
		var matchList =  includeMatchStr.match(includeReg);
		
		//如果include文件
		if(matchList && matchList[0] && matchList[1]){
			includeList.push({
				's':templateSymbol[i]['s'],
				'e':templateSymbol[i]['e'],
				'type':1,//1 表示include文件
				'val':matchList[1],
			})
			
			cb.includeFiles.push(matchList[1].split("?")[0]);
			
			include++;
			continue;
		}
		
		//如果includeid id
		var matchListId = includeMatchStr.match(includeIdReg);
		if(matchListId && matchListId[0] && matchListId[1]){
			includeList.push({
				's':templateSymbol[i]['s'],
				'e':templateSymbol[i]['e'],
				'type':2,//1 表示include文件
				'val':matchListId[1],
			})
			//将includeId存入cb的includeIds数组
			cb.includeIds.push(matchListId[1].split("?")[0]);
						
			include++;
			continue;
		}

		if(!(renderReg.test(includeMatchStr)) && !(hasFh.test(includeMatchStr))){
			includeMatchStr = includeMatchStr.slice(0, -2) + ';' + includeMatchStr.slice(-2);
			includeList.push({
					's':templateSymbol[i]['s'],
					'e':templateSymbol[i]['e'],
					'type':3,//表示正常的<% %>对，非include
					'val':includeMatchStr,
			})
		}

	}

	//如果模版不包含 include 或者 includeId
	// if(includeList.length === 0){
	// 	var notHaveInclude = false
	// 	return cb(null, tempStrList, notHaveInclude);
	// }
	if(include === 0){
		var notHaveInclude = false
		aries.assemble(tempStrList, includeList, notHaveInclude, ctx, cb);
		// return cb(null, tempStrList, notHaveInclude);
	}else{
		aries.replaceIncludeStr(tempStrList, includeList, count, ctx, cb);
	}
		
}

aries.replaceIncludeStr = function replaceIncludeStr(tempStrList, includeList, count, ctx, cb){
	var asyncWorker = [];
	
	includeList.forEach(function(incItem){
		var okCallback = true
		var dealFunc = function(callback){	
					
			if(incItem.type == 1){
				
				//去解析include的内容
				var tmpList = aries.splitIncludeParam(incItem.val);
				if(tmpList[2]){
					return callback(tmpList[2])
				}
							
				aries.getTemplateFromFile(tmpList[0], function(err, tempstr){
					if(err) return callback(err);
					incItem.template =  tmpList[1] + tempstr.toString();
					//tempStrList[0] = tempStrList[0].slice(0, incItem['s']) + tempstr + tempStrList[0].slice(incItem['e']);
					okCallback && callback();
					okCallback = false
					return
				})
			}else if(incItem.type == 2){
						
				if(!te.includeId){
					okCallback && callback(new Error("[ariestp] not defined opt.includeId can't use includeId in template!"));
					okCallback = false
					return
				}else{
					
					//去解析include的内容
					var tmpList = aries.splitIncludeParam(incItem.val);
					if(tmpList[2]){
						return callback(tmpList[2])
					}
					
					atry(function() {
					  	te.includeId(tmpList[0], function(err, tempstr){
							if(err){
								okCallback && callback(err);
								okCallback = false
								return
							}

							process.nextTick(function () {
								incItem.template = tmpList[1] + tempstr.toString();
								// tempStrList[0] = tempStrList[0].slice(0, incItem['s']) + tempstr + tempStrList[0].slice(incItem['e']);
								okCallback && callback();
								okCallback = false
							});
							return;
						});
					}).catch(function(err) {
						//定义ariestp错误
						ctx._ariestpError = err;

						okCallback && callback(err);
						okCallback = false
						return
					});

				}
			}
			else{ //type == 3
				incItem.template = incItem.val;
				okCallback && callback();
				okCallback = false
				return
			}
		}
		asyncWorker.push(dealFunc);	
		
	});
	//异步并发
	async.parallel(asyncWorker, function(err){
		if(err){
			return cb(err);
		} 
		//重新做检查，看是否有include
		var len = includeList.length;
		for(var i=len-1;i>=0;i--){

			tempStrList[0] = tempStrList[0].slice(0, includeList[i]["s"]) + includeList[i]["template"] + tempStrList[0].slice(includeList[i]["e"]);
		}
		aries.scanInclude(tempStrList,count,ctx,cb);
	})
	
}

aries.assemble = function assemble(tempStrList, includeList, notHaveInclude, ctx, cb){
	
	var len = includeList.length;
	for(var i=len-1;i>=0;i--){
			tempStrList[0] = tempStrList[0].slice(0, includeList[i]["s"]) + includeList[i]["val"] + tempStrList[0].slice(includeList[i]["e"]);
		}

	cb(null, tempStrList, notHaveInclude, cb.includeIds, cb.includeFiles);	
}

//添加功能
//include xxxxx?a=1 这样的方式传参
aries.splitIncludeParam = function(rawIncludeStr){
			
	var list = rawIncludeStr.trim().split('?');
	var includeStr = (list[0]||'').trim();
	var queryParam = (list[1]||'').trim();
	//如果没有?没有参数
	if(!queryParam){
		return [includeStr, '', null];
	}
	
	//对url query字符串进行parse解析
	try{
		var obj = url.parse('/?'+queryParam, true);
	}catch(e){
		var err = new Error(util.format('include or includeid query error: %s \n %s', rawIncludeStr, e.stack||''));
		err.stack = e.stack || '';
		return ['', '', err];
	}
	
	//添加模板字符串
	var tmpStr = '<%';
	Object.keys(obj.query).forEach(function(key){
		tmpStr += util.format(' ctx.%s = "%s"; ', key, obj.query[key]);
	})
	tmpStr += '%>';
	
	return [includeStr, tmpStr, null];
}

exports.set = aries.set;
exports.compileFile = aries.compileFile;
exports.compile = aries.compile;
exports.includes = aries.includes;
exports.splitIncludeParam = aries.splitIncludeParam;







example/exampleHttpServer.js
"use strict";
const http = require("http");
const url = require("url");
const fs = require("fs");
const aries = require("../index.js");
const path = require("path");
const util = require('util');

//设置白羊座
aries.set({
	path:__dirname,
	includeId:function(templateId, cb){
		return fs.readFile(path.join(__dirname,templateId+'.html'), cb)
	},
	cacheTime:10000,
})

let server = http.createServer((req, res) => {

    let parseObj = url.parse(req.url, true);

    if(parseObj.pathname.indexOf("/sync") === 0){
    	return syncTemplate(req, res);
    }

    if(parseObj.pathname.indexOf("/async") === 0){
    	return asyncTemplate(req, res);
    }
    if(parseObj.pathname.indexOf("/debug") === 0){
    	return debugTemplate(req, res);
    }
    if(parseObj.pathname.indexOf("/fature") === 0){
    	return templateFature(req, res);
    }
	

    	
	res.statusCode = 200;
	res.end("not found");
 
});

server.listen(3001);
console.log("listen port "+ 3001)

//异步错误响应
let errorResp = (req, res, err) => {
	res.writeHead(500, {
		"Content-Type": "text/html",
	});
    res.end(util.format("%s", err.stack));
}


//同步渲染模板
let syncTemplate = (req, res) => {
	let s1 = Date.now()
	res.end("<% includeId 5710b3053caccff63744a3f5 %>"+'<br/>消耗时间：'+(Date.now()-s1)+';')

}



//异步渲染模板
let asyncTemplate = (req, res) => {
	let s1 = Date.now()
	let ctx = {
		"http":http,
		"msg":"这时一个异步模版",
		"fs":fs,
	}
	aries.compileFile("async_example.html", ctx, (err, renderStr, isUseCache) => {
		   err && errorResp(req, res, err);
		   res.writeHead(200, {
				"Content-Type": "text/html",
			});
	       res.end(renderStr+'<br/>消耗时间：'+(Date.now()-s1)+'; 是否使用缓存：'+ (isUseCache||"false"));
	})

}
//模板支持的特性
let templateFature = (req, res) => {
	let s1 = Date.now()
	let ctx = {
		print:function(str) { 
			return util.format("%s + %s + %s", str, str, str)
		},
		data:["a","b","c","d","e","f","g"]
	};

	aries.compileFile("fature_example.html", ctx, (err, renderStr, isUseCache) => {
		   err && errorResp(req, res, err);
		   res.writeHead(200, {
				"Content-Type": "text/html",
			});
	       res.end(renderStr+'<br/>消耗时间：'+(Date.now()-s1)+'; 是否使用缓存：'+(isUseCache||"false"));
	})

}


//测试控制器
let debugTemplate = (req, res) => {
	let s1 = Date.now()
	let ctx = {
		//"http":http,
		"msg":"这时一个异步模版",
		"fs":fs,
	}
	aries.compileFile("async_debug.html", ctx, (err, renderStr, isUseCache) => {
		   err && errorResp(req, res, err);
		   res.writeHead(200, {
				"Content-Type": "text/html",
			});
	       res.end(renderStr+'<br/>消耗时间：'+(Date.now()-s1)+'; 是否使用缓存：'+ (isUseCache||"false"));
	})

}